src/emqx_session.erl

```diff
@@ -634,16 +625,44 @@ replay(ClientInfo, Session = #session{inflight = Inflight}) ->
     end.
 
 -spec(terminate(emqx_types:clientinfo(), Reason :: term(), session()) -> ok).
-terminate(ClientInfo, discarded, Session) ->
+terminate(ClientInfo, Reason, Session) ->
+    run_terminate_hooks(ClientInfo, Reason, Session),
+    cleanup_self_from_shared_subs(),
+    redispatch_shared_messages(Session),
+    ok.
+
+run_terminate_hooks(ClientInfo, discarded, Session) ->
     run_hook('session.discarded', [ClientInfo, info(Session)]);
-terminate(ClientInfo, takeovered, Session) ->
+run_terminate_hooks(ClientInfo, takeovered, Session) ->
     run_hook('session.takeovered', [ClientInfo, info(Session)]);
-terminate(ClientInfo, Reason, Session) ->
+run_terminate_hooks(ClientInfo, Reason, Session) ->
     run_hook('session.terminated', [ClientInfo, Reason, info(Session)]).
 
+redispatch_shared_messages(#session{inflight = Inflight}) ->
+    InflightList = emqx_inflight:to_list(Inflight),
+    lists:map(fun
+        %% Only QoS1 messages get redispatched, because QoS2 messages
+        %% must be sent to the same client, once they were sent
+        ({_, {#message{topic = Topic, qos = ?QOS_1} = Msg, _}}) ->
```

Comment:
A suggestion: maybe we can add some warning/error logs if a QOS2 messages is not completed